### ISSUE TITLE
ClearLinux 0122-locking-rwsem-spin-faster.patch kernel patch contain an integer overflow.

### ISSUE THREAD

**Issue by Fox2Code** [clearlinux/distribution#2296]
I found that the patch [0122-locking-rwsem-spin-faster.patch](https://github.com/clearlinux-pkgs/linux/blob/master/0122-locking-rwsem-spin-faster.patch) contain an integer overflow.

Because `if (i++ > 1000)` increase `int i` unconditionaly, the value can overflow to the negatives.

I think changing this
```.patch
+		if (i++ > 1000)
+			cpu_relax();
```
to this
```.patch
+		if (i > 1000)
+			cpu_relax();
+		else i++;
```
could resolve the issue by stoping to increment `i` when it reaches `1001`.

I reported the issue here first https://github.com/Frogging-Family/linux-tkg/issues/194 but I decided to report here just after checking that the issue also applied to the official ClearLinux kernel because I think it's more appropriate to report the issue here.

**Comment by mdhorn** [comment]
@fenrus75 Any thoughts on this patch?

**Comment by fenrus75** [comment]
it's not how I would solve it but it's also harmless either way.
(I'd solve it by making the counter unsigned)

**Comment by thiagomacieira** [comment]
The proposed solution doesn't fix the overflow: the overflow could happen before, it can still now.

Making the variable unsigned does fix the *signed* integer overflow, at the cost of not relaxing again for 1000 loops when it does overflow (note: it takes twice as long to overflow).

**Comment by fenrus75** [comment]
yes it'll eventually overflow and do another 1000 non-pause loops. That's
fine.....

(a pause is maybe 500 cycles. so it'll be a while :... and anon-pause is
very cheap.. cheaper than the branch mispredicts the "else" will end up
adding)

**Comment by marioroy** [comment]
Will there be an adverse affect if ```i``` was reset to zero?
```diff
-		cpu_relax();
+		if (i++ > 1000) {
+			cpu_relax();
+			i = 0;
+		}
```

**Comment by fenrus75** [comment]
this is the wrong thing to do.
the point of the patch is to not do cpu_relax() for the first (hot) loop,
but do it for the rest of all loops

your patch makes it effectively never do cpu_relax()



(cpu_relax() is effectively a "hyperthreading yield", e.g. tells the cpu to
give the next 1000-ish cycles exclusively to the other hyperthread, which
is fine in spin loops if you're going to spin for a long time. but if
you're going to spin for super short that's a bit on the heavy side. the
point of the patch is to do quick spins for a little bit, and see if the
total spin time is super short. once it's not, you go into the slower spins
with the yield)

if the counter overflows it does not matter at all.. (first of all if you
had been spinning that long your computer already rebooted due to the
deadlock detectors)... all you do is spin the fast way for a bit until you
go back to spinning the slow way

**Comment by Fox2Code** [comment]
Using an `unsigned int` is better, when I first made this issue I didn't see that because I was tiered.
`+	int i = 0;` -> `+	unsigned int i = 0;`
While it can still overflow, the effect of the overflow would be limited to only 1000 spins, making the impact negligible.
This new patch also avoid branching code.

New file location: [0121-locking-rwsem-spin-faster.patch](https://github.com/clearlinux-pkgs/linux/blob/master/0121-locking-rwsem-spin-faster.patch)
